server/src/main/java/io/crate/types/ArrayType.java

```diff
@@ -99,7 +100,17 @@ public ArrayType(DataType<T> innerType) {
                 public ValueIndexer<T> valueIndexer(RelationName table,
                                                     Reference ref,
                                                     Function<ColumnIdent, Reference> getRef) {
-                    return new ArrayIndexer<>(innerStorage.valueIndexer(table, ref, getRef), getRef, ref);
+                    int topMostArrayDimensions = ArrayType.dimensions(innerType) + 1;
+                    if (topMostArrayDimensions != ArrayType.dimensions(ref.valueType())) {
+                        throw new IllegalStateException("Must not retrieve value indexer of the child array of a multi dimensional array");
+                    }
```

Comment:
For example for `int_array_array`, the valueIndexer for the child array, `int_array`, should never be used in any ways.

Reply:
And the message could point towards what to do. E.g. `Must use ChildArrayIndexer for nested columns`